# HAMOYE QUIZ: STAGE A

In [1]:
import numpy as np
import pandas as pd
import pprint

In [2]:
fuel_orig = pd.read_csv('fuel_ferc1.csv')

In [3]:
fuel_orig.shape

(29523, 11)

In [4]:
fuel_orig.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

# QUESTION 1

In [5]:
a = [1, 2, 3, 4, 5, 6]
b = [13, 21, 34]

In [6]:
a.extend(b)
print(a)

[1, 2, 3, 4, 5, 6, 13, 21, 34]


# QUESTION 2

In [7]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

# QUESTION 3

In [8]:
fuel_orig.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean().sort_values(ascending=False)

fuel_type_code_pudl
nuclear    4955.157002
oil         168.877086
coal         67.421830
waste        19.518122
other        18.253856
gas          13.659397
Name: fuel_cost_per_unit_burned, dtype: float64

- The fuel type with the lowest average fuel cost per unit burned is **'Gas'**

# QUESTION 4

In [9]:
# 4
fuel_orig.describe()['fuel_mmbtu_per_unit']

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

- Standard deviation is **10.60** to 2 d.p
- 75th percentile is **17.01** to 2 d.p

# QUESTION 5

In [10]:
print('Skewness: ', fuel_orig['fuel_qty_burned'].skew())

print('Kurtosis ', fuel_orig['fuel_qty_burned'].kurtosis())

Skewness:  15.851495469109503
Kurtosis  651.3694501337732


- Skewness of the fuel quantity burned is **15.85** to 2 d.p
- Kurtosis of the fuel quantity burned is **651.37** to 2 d.p

# QUESTION 6

In [11]:
# 6
pd.isnull(fuel_orig).sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [12]:
# no of missing values is 180
# total number of values is 29523
print("percentage of missing row is '{}'".format((180/29523)*100))

percentage of missing row is '0.609694136774718'


Feature with missing value: **'fuel_unit'**

Total number of missing value: **180**

Percentage of missing row as a factor of total number of row: **0.610** to 3 d.p

# QUESTION 7

In [13]:
fuel_orig.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


Non-Null count shows that 'fuel_unit' is the feature with the missing value and the dtype is an object(which shows it is Categorical)

In [14]:
fuel_orig['fuel_unit'].value_counts(dropna=False)

mcf       11354
ton        8958
bbl        7998
gramsU      464
mmbtu       180
NaN         180
kgU         110
mwhth       100
mwdth        95
gal          84
Name: fuel_unit, dtype: int64

In [15]:
fuel_orig['fuel_unit'] = fuel_orig['fuel_unit'].fillna(value='mcf')
fuel_orig.isnull().sum()

record_id                       0
utility_id_ferc1                0
report_year                     0
plant_name_ferc1                0
fuel_type_code_pudl             0
fuel_unit                       0
fuel_qty_burned                 0
fuel_mmbtu_per_unit             0
fuel_cost_per_unit_burned       0
fuel_cost_per_unit_delivered    0
fuel_cost_per_mmbtu             0
dtype: int64

It is advisable to fill Categorical features with the most frequent values which is **'mcf'** for this feature in the dataset.

# QUESTION 8

In [16]:
fuel_orig.corr()['fuel_cost_per_unit_burned'].sort_values()

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

Features with the second and third lowest correlation with fuel cost per unit burned are;

- **'fuel_qty_burned'**
- **'fuel_mmbtu_per_unit'**

# QUESTION 9

Formula to calculate percentage change;

((new - old)/old)*100

In [17]:
fuel_orig['fuel_type_code_pudl'].value_counts()

gas        11486
coal        8547
oil         8064
nuclear      818
waste        441
other        167
Name: fuel_type_code_pudl, dtype: int64

In [18]:

cost_per_unit_burned = {}

for value in [1994, 1998]:
    condition = (fuel_orig['fuel_type_code_pudl'] == 'coal')&(fuel_orig['report_year'] == value)
    ans = fuel_orig[condition].loc[:, 'fuel_cost_per_unit_burned']
    cost_per_unit_burned[value] = sum(ans)
     
pprint.pprint(cost_per_unit_burned)

{1994: 14984.572000000011, 1998: 11902.597000000012}


In [19]:
new_year = 1998
old_year = 1994

In [20]:
((cost_per_unit_burned[new_year] - cost_per_unit_burned[old_year])/cost_per_unit_burned[old_year])*100

-20.56765451826049

Percentange change in fuel cost per unit burned in 1998 compared to 1994 is **-21%**

# QUESTION 10

In [21]:
fuel_orig.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().sort_values(ascending=False)

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64

The year with the highest average fuel cost per unit delivered is **1997**